From 608f58a55b6f39416f6b8b344154e6283808f09a Mon Sep 17 00:00:00 2001
From: gao.cui <gao.cui@windsolve.com>
Date: Sun, 13 Jul 2014 22:55:32 +0800
Subject: [PATCH 28/30] Set bit 5 in S/PDIF register SCR clear outgoing validity bit

---
 sound/soc/codecs/mxc_spdif.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/sound/soc/codecs/mxc_spdif.c b/sound/soc/codecs/mxc_spdif.c
index b358f71..54adb84 100644
--- a/sound/soc/codecs/mxc_spdif.c
+++ b/sound/soc/codecs/mxc_spdif.c
@@ -577,6 +577,7 @@ static int mxc_spdif_playback_start(struct snd_pcm_substream *substream,
 	regval &= ~SCR_LOW_POWER;
 	regval |= SCR_TXFIFO_AUTOSYNC | SCR_TXFIFO_NORMAL |
 	    SCR_TXSEL_NORMAL | SCR_USRC_SEL_CHIP | (2 << SCR_TXFIFO_ESEL_BIT);
+	regval |= SCR_VAL_CLEAR;
 	__raw_writel(regval, SPDIF_REG_SCR + spdif_base_addr);
 
 	/* Default clock source from EXTAL, divider by 8, generate 44.1kHz
-- 
1.7.0.4

