diff --git a/board/freescale/mx6q_sabresd/mx6q_sabresd.c b/board/freescale/mx6q_sabresd/mx6q_sabresd.c
index 6f866aa..bddb071 100644
--- a/board/freescale/mx6q_sabresd/mx6q_sabresd.c
+++ b/board/freescale/mx6q_sabresd/mx6q_sabresd.c
@@ -91,9 +91,8 @@ DECLARE_GLOBAL_DATA_PTR;
 
 static enum boot_device boot_dev;
 
-#define GPIO_VOL_DN_KEY IMX_GPIO_NR(1, 5)
-#define USB_OTG_PWR IMX_GPIO_NR(3, 22)
-#define USB_H1_POWER IMX_GPIO_NR(1, 29)
+#define USB_OTG_PWR IMX_GPIO_NR(4, 15)
+#define USB_H1_POWER IMX_GPIO_NR(1, 0)
 
 extern int sata_curr_device;
 
@@ -864,59 +863,6 @@ void spi_io_init(struct imx_spi_dev_t *dev)
 	}
 }
 #endif
-#if 0
-#ifdef CONFIG_NAND_GPMI
-
-iomux_v3_cfg_t nfc_pads[] = {
-	MX6Q_PAD_NANDF_CLE__RAWNAND_CLE,
-	MX6Q_PAD_NANDF_ALE__RAWNAND_ALE,
-	MX6Q_PAD_NANDF_WP_B__RAWNAND_RESETN,
-	MX6Q_PAD_NANDF_RB0__RAWNAND_READY0,
-	MX6Q_PAD_NANDF_CS0__RAWNAND_CE0N,
-	MX6Q_PAD_NANDF_CS1__RAWNAND_CE1N,
-	MX6Q_PAD_NANDF_CS2__RAWNAND_CE2N,
-	MX6Q_PAD_NANDF_CS3__RAWNAND_CE3N,
-	MX6Q_PAD_SD4_CMD__RAWNAND_RDN,
-	MX6Q_PAD_SD4_CLK__RAWNAND_WRN,
-	MX6Q_PAD_NANDF_D0__RAWNAND_D0,
-	MX6Q_PAD_NANDF_D1__RAWNAND_D1,
-	MX6Q_PAD_NANDF_D2__RAWNAND_D2,
-	MX6Q_PAD_NANDF_D3__RAWNAND_D3,
-	MX6Q_PAD_NANDF_D4__RAWNAND_D4,
-	MX6Q_PAD_NANDF_D5__RAWNAND_D5,
-	MX6Q_PAD_NANDF_D6__RAWNAND_D6,
-	MX6Q_PAD_NANDF_D7__RAWNAND_D7,
-	MX6Q_PAD_SD4_DAT0__RAWNAND_DQS,
-};
-
-int setup_gpmi_nand(void)
-{
-	unsigned int reg;
-
-	/* config gpmi nand iomux */
-	mxc_iomux_v3_setup_multiple_pads(nfc_pads,
-			ARRAY_SIZE(nfc_pads));
-
-
-	/* config gpmi and bch clock to 11Mhz*/
-	reg = readl(CCM_BASE_ADDR + CLKCTL_CS2CDR);
-	reg &= 0xF800FFFF;
-	reg |= 0x01E40000;
-	writel(reg, CCM_BASE_ADDR + CLKCTL_CS2CDR);
-
-	/* enable gpmi and bch clock gating */
-	reg = readl(CCM_BASE_ADDR + CLKCTL_CCGR4);
-	reg |= 0xFF003000;
-	writel(reg, CCM_BASE_ADDR + CLKCTL_CCGR4);
-
-	/* enable apbh clock gating */
-	reg = readl(CCM_BASE_ADDR + CLKCTL_CCGR0);
-	reg |= 0x0030;
-	writel(reg, CCM_BASE_ADDR + CLKCTL_CCGR0);
-
-}
-#endif
-#endif
 
 #ifdef CONFIG_NET_MULTI
 int board_eth_init(bd_t *bis)
@@ -1749,29 +1695,6 @@ int board_init(void)
 }
 
 
-#ifdef CONFIG_ANDROID_RECOVERY
-
-int check_recovery_cmd_file(void)
-{
-	int button_pressed = 0;
-	int recovery_mode = 0;
-
-	recovery_mode = check_and_clean_recovery_flag();
-
-	/* Check Recovery Combo Button press or not. */
-	mxc_iomux_v3_setup_pad(MX6X_IOMUX(PAD_GPIO_5__GPIO_1_5));
-
-	gpio_direction_input(GPIO_VOL_DN_KEY);
-
-	if (gpio_get_value(GPIO_VOL_DN_KEY) == 0) { /* VOL_DN key is low assert */
-		button_pressed = 1;
-		printf("Recovery key pressed\n");
-	}
-
-	return recovery_mode || button_pressed;
-}
-#endif
-
 int board_late_init(void)
 {
 	int ret = 0;
@@ -1829,6 +1752,8 @@ int mx6_rgmii_rework(char *devname, int phy_addr)
 {
 	unsigned short val;
 
+	return 0;
+
 	/* To enable AR8031 ouput a 125MHz clk from CLK_25M */
 	phy_write(devname, phy_addr, 0xd, 0x7);
 	phy_write(devname, phy_addr, 0xe, 0x8016);
@@ -2000,9 +1925,9 @@ void udc_pins_setting(void)
 	mxc_iomux_v3_setup_pad(MX6X_IOMUX(PAD_ENET_TXD1__GPIO_1_29));
 
 	/* USB_OTG_PWR = 0 */
-	gpio_direction_output(USB_OTG_PWR, 0);
+	gpio_direction_output(USB_OTG_PWR, 1);
 	/* USB_H1_POWER = 1 */
-	gpio_direction_output(USB_H1_POWER, 1);
+	gpio_direction_output(USB_H1_POWER, 0);
 
 	mxc_iomux_set_gpr_register(1, 13, 1, 0);
 
diff --git a/drivers/net/mxc_fec.c b/drivers/net/mxc_fec.c
index 212abb7..18d79f1 100644
--- a/drivers/net/mxc_fec.c
+++ b/drivers/net/mxc_fec.c
@@ -110,9 +110,12 @@ extern int mx6_rgmii_rework(char *devname, int phy_addr);
 				PHY_OUID_GET((valh & 0x3F), vall)
 #define PHY_MICREL_OUID_GET(valh, vall) \
 				PHY_OUID_GET((valh >> 10), vall)
+#define PHY_RTL8211_OUID_GET(valh, vall) \
+				PHY_OUID_GET((valh >> 10), vall)
 #define PHY_AR8031_OUID         PHY_OUID_GET(0x34, 0x4D)
 #define PHY_AR8033_OUID		PHY_OUID_GET(0x34, 0x4D)
 #define PHY_KSZ9021_OUID	PHY_OUID_GET(0x5, 0x22)
+#define PHY_RTL8211_OUID    PHY_OUID_GET(0x32, 0x1C) // 0x00A800E0
 
 /* Get phy ID */
 static unsigned int fec_phy_ouid;
@@ -122,6 +125,8 @@ static unsigned int fec_phy_ouid;
 			? 1 : 0)
 #define phy_is_ar8033()	(((fec_phy_ouid & PHY_OUID_MASK) == PHY_AR8033_OUID) \
 			? 1 : 0)
+#define phy_is_rtl8211()  (((fec_phy_ouid & PHY_OUID_MASK) == PHY_RTL8211_OUID) \
+			? 1 : 0)
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -226,11 +231,17 @@ static void mxc_get_phy_ouid(char *devname, int phy_addr)
 		return;
 	else if (phy_is_ar8033())
 		return;
+	else if (phy_is_rtl8211())
+		return;
 
 	fec_phy_ouid = PHY_MICREL_OUID_GET(value2, value1);
 	if (phy_is_ksz9021())
 		return;
 
+	fec_phy_ouid = PHY_RTL8211_OUID_GET(value2, value1);
+	if (phy_is_rtl8211())
+		return;
+
 	fec_phy_ouid = 0xFFFFFFFF;
 }
 
@@ -409,7 +420,7 @@ static void setFecDuplexSpeed(volatile fec_t *fecp, unsigned char addr,
 		}
 
 		/* for Atheros PHY */
-		if (phy_is_ar8031() || phy_is_ar8033()) {
+		if (phy_is_ar8031() || phy_is_ar8033() || phy_is_rtl8211()) {
 			ret = __fec_mii_read(fecp, addr, PHY_MIPSCR, &val);
 			if (ret)
 				dup_spd = _100BASET | (FULL << 16);
diff --git a/include/configs/mx6dl_sabresd.h b/include/configs/mx6dl_sabresd.h
index 70ed3db..29ff727 100644
--- a/include/configs/mx6dl_sabresd.h
+++ b/include/configs/mx6dl_sabresd.h
@@ -24,6 +24,8 @@
 
 #include <asm/arch/mx6.h>
 
+//#define DEBUG
+
  /* High Level Configuration Options */
 #define CONFIG_ARMV7	/* This is armv7 Cortex-A9 CPU core */
 #define CONFIG_MXC
@@ -96,8 +98,8 @@
 #define CONFIG_BOOTP_GATEWAY
 #define CONFIG_BOOTP_DNS
 
-#define CONFIG_CMD_SPI
-#define CONFIG_CMD_I2C
+//#define CONFIG_CMD_SPI
+//#define CONFIG_CMD_I2C
 #define CONFIG_CMD_IMXOTP
 
 /* Enable below configure when supporting nand */
@@ -120,24 +122,28 @@
 #define CONFIG_LOADADDR		0x10800000	/* loadaddr env var */
 #define CONFIG_RD_LOADADDR	(CONFIG_LOADADDR + 0x300000)
 
-#define	CONFIG_EXTRA_ENV_SETTINGS					\
-		"netdev=eth0\0"						\
-		"ethprime=FEC0\0"					\
-		"uboot=u-boot.bin\0"			\
-		"kernel=uImage\0"				\
-		"nfsroot=/opt/eldk/arm\0"				\
-		"bootargs_base=setenv bootargs console=ttymxc0,115200\0"\
-		"bootargs_nfs=setenv bootargs ${bootargs} root=/dev/nfs "\
-			"ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0"\
-		"bootcmd_net=run bootargs_base bootargs_nfs; "		\
-			"tftpboot ${loadaddr} ${kernel}; bootm\0"	\
-		"bootargs_mmc=setenv bootargs ${bootargs} ip=dhcp "     \
-			"root=/dev/mmcblk0p1 rootwait\0"                \
-		"bootcmd_mmc=run bootargs_base bootargs_mmc; "   \
-		"mmc dev 3; "	\
-		"mmc read ${loadaddr} 0x800 0x2000; bootm\0"	\
-		"bootcmd=run bootcmd_net\0"                             \
-
+#define	CONFIG_EXTRA_ENV_SETTINGS                                        \
+		"netdev=eth0\0"                                          \
+		"ethprime=FEC0\0"                                        \
+		"uboot=u-boot.bin\0"                                     \
+		"kernel=uImage\0"                                        \
+		"bootargs_base=console=ttymxc0,115200\0"                 \
+		"bootargs_nfs=setenv bootargs ${bootargs_base} ip=dhcp " \
+		        "root=/dev/nfs ro rootwait consoleblank=0'\0"    \
+		"bootcmd_nfs=run bootargs_nfs; dhcp ${loadaddr}; bootm\0"\
+		"bootargs_sd=setenv bootargs ${bootargs_base} ip=none "	 \
+			"root=/dev/mmcblk1p2 rootwait consoleblank=0\0"	 \
+		"bootcmd_sd=run bootargs_sd; mmc dev 2; "                \
+			"ext2load mmc 2:1 ${loadaddr} /uImage; bootm\0"	 \
+		"bootargs_mmc=setenv bootargs ${bootargs_base} ip=none " \
+			"root=/dev/mmcblk0p2 rootwait consoleblank=0\0"	 \
+		"bootcmd_mmc=run bootargs_mmc; mmc dev 3; "              \
+			"ext2load mmc 3:1 ${loadaddr} /uImage; bootm\0"	 \
+		"bootargs_rec=setenv bootargs ${bootargs_base} ip=none " \
+			"root=/dev/mmcblk0p3 rootwait consoleblank=0\0"	 \
+		"bootcmd_rec=run bootargs_rec; mmc dev 3; "              \
+			"ext2load mmc 3:1 ${loadaddr} /uImage-curie; bootm\0"	 \
+		"bootcmd=run bootcmd_sd; run bootcmd_rec; run bootcmd_mmc; run bootcmd_nfs\0"
 
 #define CONFIG_ARP_TIMEOUT	200UL
 
@@ -145,7 +148,7 @@
  * Miscellaneous configurable options
  */
 #define CONFIG_SYS_LONGHELP		/* undef to save memory */
-#define CONFIG_SYS_PROMPT		"MX6SDL SABRESD U-Boot > "
+#define CONFIG_SYS_PROMPT		"CurieSD U-Boot > "
 #define CONFIG_AUTO_COMPLETE
 #define CONFIG_SYS_CBSIZE		1024	/* Console I/O Buffer Size */
 /* Print Buffer Size */
diff --git a/include/configs/mx6q_sabresd.h b/include/configs/mx6q_sabresd.h
index 6f07961..d3409e6 100644
--- a/include/configs/mx6q_sabresd.h
+++ b/include/configs/mx6q_sabresd.h
@@ -90,8 +90,8 @@
 #define CONFIG_BOOTP_GATEWAY
 #define CONFIG_BOOTP_DNS
 
-#define CONFIG_CMD_SPI
-#define CONFIG_CMD_I2C
+//#define CONFIG_CMD_SPI
+//#define CONFIG_CMD_I2C
 #define CONFIG_CMD_IMXOTP
 
 /* Enable below configure when supporting nand */
@@ -115,24 +115,28 @@
 #define CONFIG_LOADADDR		0x10800000	/* loadaddr env var */
 #define CONFIG_RD_LOADADDR	(0x1300000)
 
-#define	CONFIG_EXTRA_ENV_SETTINGS					\
-		"netdev=eth0\0"						\
-		"ethprime=FEC0\0"					\
-		"uboot=u-boot.bin\0"			\
-		"kernel=uImage\0"				\
-		"nfsroot=/opt/eldk/arm\0"				\
-		"bootargs_base=setenv bootargs console=ttymxc0,115200\0"\
-		"bootargs_nfs=setenv bootargs ${bootargs} root=/dev/nfs "\
-			"ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0"\
-		"bootcmd_net=run bootargs_base bootargs_nfs; "		\
-			"tftpboot ${loadaddr} ${kernel}; bootm\0"	\
-		"bootargs_mmc=setenv bootargs ${bootargs} ip=dhcp "     \
-			"root=/dev/mmcblk0p1 rootwait\0"                \
-		"bootcmd_mmc=run bootargs_base bootargs_mmc; "   \
-		"mmc dev 3; "	\
-		"mmc read ${loadaddr} 0x800 0x2000; bootm\0"	\
-		"bootcmd=run bootcmd_net\0"                             \
-
+#define	CONFIG_EXTRA_ENV_SETTINGS					 \
+		"netdev=eth0\0"                                          \
+		"ethprime=FEC0\0"                                        \
+		"uboot=u-boot.bin\0"                                     \
+		"kernel=uImage\0"                                        \
+		"bootargs_base=console=ttymxc0,115200\0"                 \
+		"bootargs_nfs=setenv bootargs ${bootargs_base} ip=dhcp " \
+		        "root=/dev/nfs ro rootwait consoleblank=0'\0"    \
+		"bootcmd_nfs=run bootargs_nfs; dhcp ${loadaddr}; bootm\0"\
+		"bootargs_sd=setenv bootargs ${bootargs_base} ip=none "	 \
+			"root=/dev/mmcblk1p2 rootwait consoleblank=0\0"	 \
+		"bootcmd_sd=run bootargs_sd; mmc dev 2; "                \
+			"ext2load mmc 2:1 ${loadaddr} /uImage; bootm\0"	 \
+		"bootargs_mmc=setenv bootargs ${bootargs_base} ip=none " \
+			"root=/dev/mmcblk0p2 rootwait consoleblank=0\0"	 \
+		"bootcmd_mmc=run bootargs_mmc; mmc dev 3; "              \
+			"ext2load mmc 3:1 ${loadaddr} /uImage; bootm\0"	 \
+		"bootargs_rec=setenv bootargs ${bootargs_base} ip=none " \
+			"root=/dev/mmcblk0p3 rootwait consoleblank=0\0"	 \
+		"bootcmd_rec=run bootargs_rec; mmc dev 3; "              \
+			"ext2load mmc 3:1 ${loadaddr} /uImage-curie; bootm\0"	 \
+		"bootcmd=run bootcmd_sd; run bootcmd_rec; run bootcmd_mmc; run bootcmd_nfs\0"
 
 #define CONFIG_ARP_TIMEOUT	200UL
 
@@ -140,7 +141,7 @@
  * Miscellaneous configurable options
  */
 #define CONFIG_SYS_LONGHELP		/* undef to save memory */
-#define CONFIG_SYS_PROMPT		"MX6Q SABRESD U-Boot > "
+#define CONFIG_SYS_PROMPT		"CurieQ U-Boot > "
 #define CONFIG_AUTO_COMPLETE
 #define CONFIG_SYS_CBSIZE		1024	/* Console I/O Buffer Size */
 /* Print Buffer Size */
@@ -171,8 +172,8 @@
 #define CONFIG_CMD_MII
 #define CONFIG_CMD_DHCP
 #define CONFIG_CMD_PING
-#define CONFIG_IPADDR			192.168.1.103
-#define CONFIG_SERVERIP			192.168.1.101
+#define CONFIG_IPADDR			192.168.18.223
+#define CONFIG_SERVERIP			192.168.18.123
 #define CONFIG_NETMASK			255.255.255.0
 
 /*
