diff --git a/psplash-config.h b/psplash-config.h
index 82bb76d..a391222 100644
--- a/psplash-config.h
+++ b/psplash-config.h
@@ -20,15 +20,23 @@
 #define _HAVE_PSPLASH_CONFIG_H
 
 /* Text to output on program start; if undefined, output nothing */
+#ifndef PSPLASH_STARTUP_MSG
 #define PSPLASH_STARTUP_MSG ""
+#endif
 
 /* Bool indicating if the image is fullscreen, as opposed to split screen */
+#ifndef PSPLASH_IMG_FULLSCREEN
 #define PSPLASH_IMG_FULLSCREEN 0
+#endif
 
 /* Position of the image split from top edge, numerator of fraction */
+#ifndef PSPLASH_IMG_SPLIT_NUMERATOR
 #define PSPLASH_IMG_SPLIT_NUMERATOR 5
+#endif
 
 /* Position of the image split from top edge, denominator of fraction */
+#ifndef PSPLASH_IMG_SPLIT_DENOMINATOR
 #define PSPLASH_IMG_SPLIT_DENOMINATOR 6
+#endif
 
 #endif
