From 06acc4799a3acf793c86a53f5ad31baa6432ce04 Mon Sep 17 00:00:00 2001
From: Cui Gao <gao.cui@windsolve.com>
Date: Tue, 5 Aug 2014 13:09:03 +0800
Subject: [PATCH 3/3] add imx6 duallite curie board

---
 boards.cfg | 1 +
 1 file changed, 1 insertion(+)

diff --git a/boards.cfg b/boards.cfg
index de20e11..3a53da5 100644
--- a/boards.cfg
+++ b/boards.cfg
@@ -278,6 +278,7 @@ mx6qsabresd                  arm         armv7       mx6qsabresd         freesca
 mx6qsabresd_sata             arm         armv7       mx6qsabresd         freescale      mx6		mx6qsabresd:IMX_CONFIG=board/freescale/imx/ddr/mx6q_4x_mt41j128.cfg,MX6Q,DEFAULT_FDT_FILE="imx6q-sabresd.dtb",DDR_MB=1024,SYS_USE_SPINOR,SYS_BOOT_SATA
 mx6qsabresdandroid       arm         armv7       mx6qsabresd         freescale      mx6		mx6qsabresd:IMX_CONFIG=board/freescale/imx/ddr/mx6q_4x_mt41j128.cfg,MX6Q,DEFAULT_FDT_FILE="imx6q-sabresd.dtb",DDR_MB=1024,SYS_USE_SPINOR,ANDROID_SUPPORT
 mx6qcurie                  arm         armv7       mx6qcurie         freescale      mx6		mx6qcurie:IMX_CONFIG=board/freescale/imx/ddr/mx6q_4x_mt41j128.cfg,MX6Q,DEFAULT_FDT_FILE="imx6q-curie.dtb",DDR_MB=1024,SYS_USE_SPINOR
+mx6dlcurie                 arm         armv7       mx6qcurie         freescale      mx6		mx6qcurie:IMX_CONFIG=board/freescale/mx6qsabresd/mx6dl_4x_mt41j128.cfg,MX6DL,DEFAULT_FDT_FILE="imx6dl-curie.dtb",DDR_MB=1024,SYS_USE_SPINOR
 mx6solosabresd               arm         armv7       mx6qsabresd         freescale      mx6             mx6qsabresd:IMX_CONFIG=board/freescale/mx6qsabresd/mx6solo_4x_mt41j128.cfg,MX6SOLO,DEFAULT_FDT_FILE="imx6dl-sabresd.dtb",DDR_MB=512,SYS_USE_SPINOR,SYS_NOSMP="nosmp"
 mx6solosabresdandroid               arm         armv7       mx6qsabresd         freescale      mx6             mx6qsabresd:IMX_CONFIG=board/freescale/mx6qsabresd/mx6solo_4x_mt41j128.cfg,MX6SOLO,DEFAULT_FDT_FILE="imx6dl-sabresd.dtb",DDR_MB=512,SYS_USE_SPINOR,SYS_NOSMP="nosmp",ANDROID_SUPPORT
 mx6solosabreauto             arm         armv7       mx6qsabreauto       freescale      mx6		mx6qsabreauto:IMX_CONFIG=board/freescale/mx6qsabreauto/mx6solo.cfg,MX6SOLO,DEFAULT_FDT_FILE="imx6dl-sabreauto.dtb",DDR_MB=1024,SYS_NOSMP="nosmp"
-- 
1.9.1

